File: SerialPortDemo/trunk/VB/Serial/Invoker.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Text

Public Class Invoker

    Public Shared Sub Invoke(ByVal delegateToCall As [Delegate], ByVal argsOfDelegateToCall As Object())

        Dim del As [Delegate]

        For Each del In delegateToCall.GetInvocationList()

            Dim si As ISynchronizeInvoke = CType(del.Target.Target, ISynchronizeInvoke)

            If si IsNot Nothing AndAlso si.InvokeRequired Then
                si.Invoke(del, argsOfDelegateToCall)
            Else
                del.DynamicInvoke(argsOfDelegateToCall)
            End If

        Next

    End Sub

End Class

File: SerialPortDemo/trunk/CS/Serial/Invoker.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Text;

namespace Serial
{
    public static class Invoker
    {
        public static void Invoke(Delegate delegateToCall, params object[] argsOfDelegateToCall)
        {
            foreach (Delegate del in delegateToCall.GetInvocationList())
            {
                ISynchronizeInvoke si = del.Target as ISynchronizeInvoke;
                if (si != null && si.InvokeRequired)
                    si.Invoke(del, argsOfDelegateToCall);
                else
                    del.DynamicInvoke(argsOfDelegateToCall);
            }
        }
    }
}